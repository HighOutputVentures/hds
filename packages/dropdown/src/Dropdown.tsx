import {
  chakra,
  Divider,
  Menu,
  MenuButton,
  MenuGroup,
  MenuItem,
  MenuList,
  PlacementWithLogical,
  Portal,
  useDisclosure,
  UseDisclosureReturn,
} from "@chakra-ui/react";
import * as React from "react";
import { v4 as uuid } from "uuid";
import { useStyles } from "./hooks";

interface Item {
  icon?: JSX.Element;
  label: React.ReactNode;
  command?: string;
  onClick?(): void;
}

interface BaseProps {
  placement?: PlacementWithLogical | undefined;
  closeOnSelect?: boolean;
  renderHeader?: JSX.Element;
  renderOption?(item: Item): JSX.Element;
  children(context: UseDisclosureReturn): JSX.Element;
  __menuTestId?: string;
  __menuItemTestId?: string;
}

interface GroupProps {
  items: Item[][];
  isGrouped: true;
}

interface SingleProps {
  items: Item[];
  isGrouped?: false;
}

export type DropdownProps = (GroupProps | SingleProps) & BaseProps;

const Dropdown: React.FC<DropdownProps> = (props) => {
  const {
    children,
    placement,
    renderHeader,
    renderOption = (item) => item.label,
    closeOnSelect,
    __menuTestId = "hds.dropdown",
    __menuItemTestId = "hds.dropdown.item",
    ...others
  } = props;

  const styles = useStyles();
  const disclosure = useDisclosure();

  return (
    <Menu
      variant="unstyled"
      isOpen={disclosure.isOpen}
      onOpen={disclosure.onOpen}
      onClose={disclosure.onClose}
      placement={placement}
      closeOnBlur
      closeOnSelect={closeOnSelect}
      isLazy
      lazyBehavior="keepMounted"
    >
      <MenuButton as={createRefReciever(children(disclosure))} />

      <Portal appendToParentPortal={false}>
        <MenuList sx={styles.menulist} data-testid={__menuTestId}>
          {!!renderHeader && (
            <MenuGroup sx={styles.menugroup}>
              <chakra.div sx={styles.header}>{renderHeader}</chakra.div>
              <Divider sx={styles.divider} />
            </MenuGroup>
          )}

          {!others.isGrouped && (
            <MenuGroup sx={styles.menugroup}>
              {others.items.map((item) => {
                return (
                  <MenuItem
                    sx={styles.menuitem}
                    key={uuid()}
                    onClick={item.onClick}
                    /**
                     *
                     * Only render icon and command if
                     * user did not provide the "renderOption" prop
                     *
                     */
                    {...(!props.renderOption && {
                      icon: item.icon,
                      command: item.command,
                    })}
                    data-testid={__menuItemTestId}
                  >
                    {renderOption(item)}
                  </MenuItem>
                );
              })}
            </MenuGroup>
          )}

          {props.isGrouped &&
            props.items.map((items, idx_0, arr_0) => {
              return (
                <MenuGroup key={uuid()} sx={styles.menugroup}>
                  {items.map((item, idx_1, arr_1) => {
                    const hasDivider =
                      idx_0 < arr_0.length - 1 && idx_1 === arr_1.length - 1;

                    return (
                      <React.Fragment key={uuid()}>
                        <MenuItem
                          icon={item.icon}
                          sx={styles.menuitem}
                          data-testid={__menuItemTestId}
                          /**
                           *
                           * See comment above
                           *
                           */
                          {...(!props.renderOption && {
                            icon: item.icon,
                            command: item.command,
                          })}
                        >
                          {renderOption(item)}
                        </MenuItem>

                        {hasDivider && <Divider sx={styles.divider} />}
                      </React.Fragment>
                    );
                  })}
                </MenuGroup>
              );
            })}
        </MenuList>
      </Portal>
    </Menu>
  );
};

function createRefReciever(children: JSX.Element) {
  return React.forwardRef<any>(function Tunnel(_, ref) {
    return React.cloneElement(children, { ref });
  });
}

export default Dropdown;
