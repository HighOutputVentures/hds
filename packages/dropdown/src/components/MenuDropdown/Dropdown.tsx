import { ChevronDownIcon, ChevronUpIcon } from '@chakra-ui/icons';
import {
  Box,
  Button,
  Divider,
  Menu,
  MenuButton,
  MenuGroup,
  MenuItem,
  MenuList,
  PlacementWithLogical,
  UseDisclosureReturn,
  useDisclosure,
} from '@chakra-ui/react';
import { ButtonVariantsTypes } from '@highoutput/hds/src/theme/components/button/variants';
import React, { FC } from 'react';

export interface Items {
  label: string;
  command?: string;
  icon?: JSX.Element;
}
export interface IMenuItemsSingle {
  items: Array<Items>;
}

export interface IMenuItemsGroup {
  items: Array<Array<Items>>;
}
export interface DropdownFieldMainProps {
  menuButtonText?: string;
  profileUrl?: string;
  indicator?: boolean;
  renderHeader?: React.ReactNode;
  __menuTestId?: string;
  __menuButtonTestId?: string;
  __menuListTestId?: string;
  gap?: string;
  placement?: PlacementWithLogical | undefined;
  variant?: ButtonVariantsTypes;
  closeOnSelect?: boolean;
  showDropdownIndicator?: boolean;
  children(context: UseDisclosureReturn): JSX.Element | null;
}
export interface DropdownFieldPropsGroup extends DropdownFieldMainProps {
  group: true;
  items: IMenuItemsGroup['items'];
  hasDivider?: never;
}
export interface DropdownFieldPropsSingle extends DropdownFieldMainProps {
  group: false;
  items: IMenuItemsSingle['items'];
  hasDivider?: boolean;
}

export type DropdownFieldProps =
  | DropdownFieldPropsGroup
  | DropdownFieldPropsSingle;
const Dropdown: FC<DropdownFieldProps> = (props) => {
  const {
    renderHeader,
    closeOnSelect = true,
    showDropdownIndicator = false,
    items,
    gap,
    group,
    hasDivider,
    children,
    __menuTestId,
    __menuButtonTestId,
    __menuListTestId,
    placement,
    variant = 'unstyled',
  } = props;
  const disclosure = useDisclosure();
  return (
    <Menu
      data-testid={__menuTestId ?? 'hds.menu.dropdown'}
      placement={placement}
      closeOnSelect={closeOnSelect}
    >
      <Box w="full">
        <MenuButton
          variant={variant}
          as={Button}
          isActive={disclosure.isOpen}
          rightIcon={
            !disclosure.isOpen && showDropdownIndicator ? (
              <ChevronDownIcon width="20px" height="20px" />
            ) : disclosure.isOpen && showDropdownIndicator ? (
              <ChevronUpIcon width="20px" height="20px" />
            ) : undefined
          }
          data-testid={__menuButtonTestId}
          display="flex"
        >
          {children(disclosure)}
        </MenuButton>

        <MenuList
          boxShadow={
            '0px 12px 16px -4px rgba(16, 24, 40, 0.08), 0px 4px 6px -2px rgba(16, 24, 40, 0.03)'
          }
          marginTop={gap}
          data-testid={__menuListTestId ?? 'hds.menu.list'}
          py={'0px'}
          fontSize={'14px'}
          color="neutrals.900"
        >
          {renderHeader}
          <Divider orientation="horizontal" />
          <MenuGroup>
            {!group &&
              (items as IMenuItemsSingle['items']).map(
                ({ command, label, icon }, idx) => {
                  return (
                    <>
                      <MenuItem command={command} icon={icon}>
                        {label}
                      </MenuItem>
                      {hasDivider && idx !== items.length - 1 && <Divider />}
                    </>
                  );
                }
              )}
            {group &&
              (items as IMenuItemsGroup['items']).map((item) => {
                return item.map(({ label, command, icon }, idx) => {
                  return (
                    <>
                      <MenuItem command={command} icon={icon}>
                        {label}
                      </MenuItem>
                      {idx === item.length - 1 && <Divider />}
                    </>
                  );
                });
              })}
          </MenuGroup>
        </MenuList>
      </Box>
    </Menu>
  );
};

export default Dropdown;
