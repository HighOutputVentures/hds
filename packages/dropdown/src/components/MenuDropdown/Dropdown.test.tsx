import { Box, ChakraProvider } from '@chakra-ui/react';
import { SettingIcon, UserIcon, ZapIcon } from '@highoutput/hds-icons';
import { render, screen } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import Dropdown, { IMenuItemsSingle } from './Dropdown';
const itemsSingle: IMenuItemsSingle['items'] = [
  { label: 'View Profile', command: '⌘K->P', icon: <UserIcon /> },
  { label: 'Settings', command: '⌘S', icon: <SettingIcon /> },
  { label: 'Keyboard shortcuts', command: '?', icon: <ZapIcon /> },
];
describe('Menu Dropdown Kebab Type Component', () => {
  beforeEach(() => {
    render(
      <ChakraProvider>
        <Dropdown items={itemsSingle} group={false} __menuButtonTestId="hds.dropdown">
          {({ onToggle }) => {
            return <Box onClick={onToggle} />;
          }}
        </Dropdown>
      </ChakraProvider>,
    );
  });

  it('Should render menu dropdown', async () => {
    const MenuKebab = await screen.findAllByTestId('hds.dropdown');

    expect(MenuKebab).toHaveLength(1);
  });
});

describe('Snapshot', () => {
  const component = renderer.create(
    <ChakraProvider>
      <Dropdown items={itemsSingle} group={false}>
        {({ onToggle }) => {
          return <Box onClick={onToggle} />;
        }}
      </Dropdown>
    </ChakraProvider>,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  component.unmount();
});
