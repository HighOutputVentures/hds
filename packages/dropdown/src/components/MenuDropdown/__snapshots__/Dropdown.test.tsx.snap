// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  <div
    className="css-1ki54i"
  >
    <button
      aria-controls="menu-list-:r2:"
      aria-expanded={false}
      aria-haspopup="menu"
      className="chakra-button chakra-menu__menu-button css-1d9ngs"
      id="menu-button-:r2:"
      onClick={[Function]}
      onKeyDown={[Function]}
      type="button"
    >
      <span
        className="css-xl71ch"
      >
        <div
          className="css-0"
          onClick={[Function]}
        />
      </span>
    </button>
    <div
      className="css-r6z5ec"
      style={
        Object {
          "inset": "0 auto auto 0",
          "minWidth": "max-content",
          "position": "absolute",
          "visibility": "hidden",
        }
      }
    >
      <div
        aria-orientation="vertical"
        className="chakra-menu__menu-list css-1xvrq0b"
        data-testid="hds.menu.list"
        id="menu-list-:r2:"
        onKeyDown={[Function]}
        role="menu"
        style={
          Object {
            "opacity": 0,
            "transform": "scale(0.8) translateZ(0)",
            "transformOrigin": "var(--popper-transform-origin)",
            "visibility": "hidden",
          }
        }
        tabIndex={-1}
      >
        <hr
          aria-orientation="horizontal"
          className="chakra-divider css-svjswr"
        />
        <div
          className="chakra-menu__group"
          role="group"
        >
          <button
            className="chakra-menu__menuitem css-18esm8n"
            id="menu-list-:r2:-menuitem-:r3:"
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseMove={[Function]}
            role="menuitem"
            tabIndex={0}
            type="button"
          >
            <span
              className="chakra-menu__icon-wrapper css-15v20fh"
            >
              <svg
                aria-hidden={true}
                className="chakra-icon chakra-menu__icon css-iuxzun"
                focusable="false"
                viewBox="0 0 16 16"
              >
                <path
                  d="M13.3334 14C13.3334 13.0696 13.3334 12.6044 13.2186 12.2259C12.9601 11.3736 12.2931 10.7067 11.4408 10.4482C11.0623 10.3333 10.5971 10.3333 9.66675 10.3333H6.33342C5.40304 10.3333 4.93785 10.3333 4.55932 10.4482C3.70705 10.7067 3.04011 11.3736 2.78157 12.2259C2.66675 12.6044 2.66675 13.0696 2.66675 14M11.0001 5C11.0001 6.65685 9.65694 8 8.00008 8C6.34323 8 5.00008 6.65685 5.00008 5C5.00008 3.34315 6.34323 2 8.00008 2C9.65694 2 11.0001 3.34315 11.0001 5Z"
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="1.5"
                />
              </svg>
            </span>
            <span
              style={
                Object {
                  "flex": 1,
                  "pointerEvents": "none",
                }
              }
            >
              View Profile
            </span>
            <span
              className="chakra-menu__command css-dt0657"
            >
              ⌘K-&gt;P
            </span>
          </button>
          <button
            className="chakra-menu__menuitem css-18esm8n"
            id="menu-list-:r2:-menuitem-:r4:"
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseMove={[Function]}
            role="menuitem"
            tabIndex={0}
            type="button"
          >
            <span
              className="chakra-menu__icon-wrapper css-15v20fh"
            >
              <svg
                aria-hidden={true}
                className="chakra-icon chakra-menu__icon css-iuxzun"
                focusable="false"
                viewBox="0 0 16 16"
              >
                <path
                  d="M7.99992 10C9.10449 10 9.99992 9.10461 9.99992 8.00004C9.99992 6.89547 9.10449 6.00004 7.99992 6.00004C6.89535 6.00004 5.99992 6.89547 5.99992 8.00004C5.99992 9.10461 6.89535 10 7.99992 10Z"
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="1.5"
                />
                <path
                  d="M12.4848 9.81822C12.4041 10.001 12.38 10.2038 12.4157 10.4004C12.4513 10.597 12.545 10.7784 12.6848 10.9213L12.7211 10.9576C12.8338 11.0702 12.9232 11.2039 12.9842 11.351C13.0452 11.4982 13.0766 11.6559 13.0766 11.8152C13.0766 11.9745 13.0452 12.1322 12.9842 12.2794C12.9232 12.4265 12.8338 12.5602 12.7211 12.6728C12.6086 12.7855 12.4749 12.8749 12.3277 12.9359C12.1806 12.9969 12.0228 13.0283 11.8636 13.0283C11.7043 13.0283 11.5465 12.9969 11.3994 12.9359C11.2522 12.8749 11.1186 12.7855 11.006 12.6728L10.9696 12.6364C10.8268 12.4967 10.6454 12.403 10.4488 12.3673C10.2522 12.3317 10.0494 12.3557 9.86659 12.4364C9.68733 12.5132 9.53445 12.6408 9.42677 12.8034C9.31909 12.966 9.2613 13.1565 9.26052 13.3516V13.4546C9.26052 13.7761 9.13282 14.0844 8.9055 14.3117C8.67819 14.539 8.36988 14.6667 8.0484 14.6667C7.72693 14.6667 7.41862 14.539 7.1913 14.3117C6.96399 14.0844 6.83628 13.7761 6.83628 13.4546V13.4C6.83159 13.1994 6.76666 13.0049 6.64993 12.8417C6.53319 12.6785 6.37006 12.5542 6.18174 12.4849C5.99894 12.4042 5.79617 12.3801 5.59956 12.4158C5.40296 12.4514 5.22154 12.5452 5.07871 12.6849L5.04234 12.7213C4.92977 12.834 4.79609 12.9234 4.64894 12.9844C4.50179 13.0454 4.34406 13.0768 4.18477 13.0768C4.02548 13.0768 3.86775 13.0454 3.7206 12.9844C3.57345 12.9234 3.43976 12.834 3.32719 12.7213C3.21449 12.6087 3.12509 12.475 3.06409 12.3278C3.00309 12.1807 2.97169 12.023 2.97169 11.8637C2.97169 11.7044 3.00309 11.5467 3.06409 11.3995C3.12509 11.2524 3.21449 11.1187 3.32719 11.0061L3.36356 10.9697C3.50327 10.8269 3.597 10.6455 3.63265 10.4489C3.6683 10.2523 3.64423 10.0495 3.56355 9.86671C3.48673 9.68745 3.35916 9.53458 3.19656 9.42689C3.03396 9.31921 2.84343 9.26142 2.6484 9.26065H2.54537C2.2239 9.26065 1.91559 9.13294 1.68827 8.90562C1.46096 8.67831 1.33325 8.37 1.33325 8.04853C1.33325 7.72705 1.46096 7.41874 1.68827 7.19143C1.91559 6.96411 2.2239 6.8364 2.54537 6.8364H2.59992C2.80052 6.83171 2.99507 6.76678 3.15828 6.65005C3.32149 6.53332 3.44581 6.37018 3.51507 6.18186C3.59575 5.99906 3.61981 5.79629 3.58416 5.59968C3.54852 5.40308 3.45479 5.22166 3.31507 5.07883L3.27871 5.04247C3.16601 4.92989 3.0766 4.79621 3.0156 4.64906C2.9546 4.50191 2.92321 4.34418 2.92321 4.18489C2.92321 4.0256 2.9546 3.86787 3.0156 3.72072C3.0766 3.57357 3.16601 3.43989 3.27871 3.32731C3.39128 3.21461 3.52496 3.12521 3.67211 3.06421C3.81926 3.00321 3.97699 2.97181 4.13628 2.97181C4.29557 2.97181 4.4533 3.00321 4.60045 3.06421C4.7476 3.12521 4.88128 3.21461 4.99386 3.32731L5.03022 3.36368C5.17306 3.5034 5.35447 3.59712 5.55108 3.63277C5.74768 3.66842 5.95045 3.64435 6.13325 3.56368H6.18174C6.36099 3.48685 6.51387 3.35929 6.62155 3.19669C6.72923 3.03409 6.78702 2.84355 6.7878 2.64853V2.5455C6.7878 2.22402 6.9155 1.91571 7.14282 1.6884C7.37014 1.46108 7.67844 1.33337 7.99992 1.33337C8.32139 1.33337 8.6297 1.46108 8.85702 1.6884C9.08433 1.91571 9.21204 2.22402 9.21204 2.5455V2.60004C9.21282 2.79506 9.27061 2.9856 9.37829 3.1482C9.48597 3.3108 9.63885 3.43837 9.8181 3.51519C10.0009 3.59587 10.2037 3.61993 10.4003 3.58429C10.5969 3.54864 10.7783 3.45491 10.9211 3.31519L10.9575 3.27883C11.0701 3.16613 11.2038 3.07673 11.3509 3.01573C11.4981 2.95473 11.6558 2.92333 11.8151 2.92333C11.9744 2.92333 12.1321 2.95473 12.2792 3.01573C12.4264 3.07673 12.5601 3.16613 12.6726 3.27883C12.7853 3.3914 12.8748 3.52509 12.9357 3.67223C12.9967 3.81938 13.0281 3.97711 13.0281 4.1364C13.0281 4.2957 12.9967 4.45343 12.9357 4.60057C12.8748 4.74772 12.7853 4.88141 12.6726 4.99398L12.6363 5.03034C12.4966 5.17318 12.4028 5.3546 12.3672 5.5512C12.3315 5.7478 12.3556 5.95058 12.4363 6.13337V6.18186C12.5131 6.36111 12.6407 6.51399 12.8033 6.62167C12.9659 6.72936 13.1564 6.78714 13.3514 6.78792H13.4545C13.7759 6.78792 14.0842 6.91563 14.3116 7.14294C14.5389 7.37026 14.6666 7.67857 14.6666 8.00004C14.6666 8.32152 14.5389 8.62982 14.3116 8.85714C14.0842 9.08446 13.7759 9.21216 13.4545 9.21216H13.3999C13.2049 9.21294 13.0144 9.27073 12.8518 9.37841C12.6892 9.48609 12.5616 9.63897 12.4848 9.81822Z"
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="1.5"
                />
              </svg>
            </span>
            <span
              style={
                Object {
                  "flex": 1,
                  "pointerEvents": "none",
                }
              }
            >
              Settings
            </span>
            <span
              className="chakra-menu__command css-dt0657"
            >
              ⌘S
            </span>
          </button>
          <button
            className="chakra-menu__menuitem css-18esm8n"
            id="menu-list-:r2:-menuitem-:r5:"
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseMove={[Function]}
            role="menuitem"
            tabIndex={0}
            type="button"
          >
            <span
              className="chakra-menu__icon-wrapper css-15v20fh"
            >
              <svg
                aria-hidden={true}
                className="chakra-icon chakra-menu__icon css-iuxzun"
                focusable="false"
                viewBox="0 0 16 16"
              >
                <path
                  d="M8.66674 1.33337L2.72904 8.45861C2.4965 8.73766 2.38023 8.87718 2.37846 8.99502C2.37691 9.09745 2.42256 9.19491 2.50224 9.2593C2.5939 9.33337 2.77552 9.33337 3.13876 9.33337H8.00007L7.3334 14.6667L13.2711 7.54147C13.5036 7.26243 13.6199 7.1229 13.6217 7.00507C13.6232 6.90263 13.5776 6.80517 13.4979 6.74078C13.4062 6.66671 13.2246 6.66671 12.8614 6.66671H8.00007L8.66674 1.33337Z"
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="1.5"
                />
              </svg>
            </span>
            <span
              style={
                Object {
                  "flex": 1,
                  "pointerEvents": "none",
                }
              }
            >
              Keyboard shortcuts
            </span>
            <span
              className="chakra-menu__command css-dt0657"
            >
              ?
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>,
  <span />,
]
`;
